runtimes/frameless-runtime/src/lib.rs

```diff
@@ -0,0 +1,301 @@
+//! A dead simple runtime that has a single boolean storage value and three transactions. The transactions
+//! available are Set, Clear, and Toggle.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+// Make the WASM binary available.
+#[cfg(feature = "std")]
+include!(concat!(env!("OUT_DIR"), "/wasm_binary.rs"));
+
+use parity_scale_codec::{Decode, Encode};
+
+use sp_std::if_std;
+use sp_std::prelude::*;
+use sp_api::impl_runtime_apis;
+use sp_runtime::{
+	ApplyExtrinsicResult,
+	create_runtime_str,
+	generic,
+	MultiSignature,
+	transaction_validity::{
+		TransactionLongevity,
+		TransactionSource,
+		TransactionValidity,
+		ValidTransaction
+	},
+	traits::{
+		BlakeTwo256,
+		Block as BlockT,
+		Extrinsic,
+		IdentifyAccount,
+		Verify,
+	}
+};
+// This strange-looking import is usually done by the `construct_runtime!` macro
+use sp_block_builder::runtime_decl_for_BlockBuilder::BlockBuilder;
+#[cfg(feature = "std")]
+use sp_storage::well_known_keys;
+
+#[cfg(any(feature = "std", test))]
+use sp_runtime::{BuildStorage, Storage};
+
+use sp_core::OpaqueMetadata;
+
+#[cfg(feature = "std")]
+use sp_version::NativeVersion;
+use sp_version::RuntimeVersion;
+
+#[cfg(feature = "std")]
+use serde::{Deserialize, Serialize};
+
+// Include the genesis helper module when building to std
+#[cfg(feature = "std")]
+pub mod genesis;
+
+/// An index to a block.
+pub type BlockNumber = u32;
+
+/// Alias to 512-bit hash when used in the context of a transaction signature on the chain.
+/// This is not currently used in the runtime but is exposed for compatability with outer
+/// nodes that expect it.
+pub type Signature = MultiSignature;
+
+/// Some way of identifying an account on the chain. We intentionally make it equivalent
+/// to the public key of our transaction signing scheme.
+/// This is not currently used in the runtime but is exposed for compatability with outer
+/// nodes that expect it.
+pub type AccountId = <<Signature as Verify>::Signer as IdentifyAccount>::AccountId;
+
+/// Opaque types. These are used by the CLI to instantiate machinery that don't need to know
+/// the specifics of the runtime. They can then be made to be agnostic over specific formats
+/// of data like extrinsics, allowing for them to continue syncing the network through upgrades
+/// to even the core datastructures.
+pub mod opaque {
+	use super::*;
+	pub use sp_runtime::OpaqueExtrinsic;
+
+	/// Opaque block header type.
+	pub type Header = generic::Header<BlockNumber, BlakeTwo256>;
+	/// Opaque block type.
+	pub type Block = generic::Block<Header, OpaqueExtrinsic>;
```

Comment:
@JoshOrndorff you are using here the `OpaqueExtrinsic` and that works a little bit different on the node side. AFAIK it has the length prepended, which you not do.

Reply:
So `OpaqueExtrinsic` is just an alias for `Vec<u8>`. And [scale docs for `Vec<_>`](https://substrate.dev/docs/en/knowledgebase/advanced/codec#vectors-lists-series-sets) do indeed say the encoding is "prefixed with a compact encoding of the number of items, followed by each item's encoding concatenated in turn".

I will be encoding a single byte which means I should prepend with the compact encoding for the number one which is `0x04` according to the same scale docs (in addition to doing the encoding myself, I notices that the number 1 is an example given.

Trying to encode my transaction this way I get the following output
```
$ curl http://localhost:9933 -H "Content-Type:application/json;charset=utf-8" -d   '{
     "jsonrpc":"2.0",
      "id":1,
      "method":"author_submitExtrinsic",
      "params": ["0x0400"]
    }'
{"jsonrpc":"2.0","error":{"code":1002,"message":"Verification Error: Execution(ApiError(\"Could not convert parameter `tx` between node and runtime: No such variant in enum FramelessTransaction\"))","data":"RuntimeApi(\"Execution(ApiError(\\\"Could not convert parameter `tx` between node and runtime: No such variant in enum FramelessTransaction\\\"))\")"},"id":1}
```

Although the error message is long, I think this is actually a step in the right direction. It is now clearly trying to decode into `FramelessTransaction` which is exactly what I want. So the remaining mysteries are:

1. Now that I know to prepend the length, how do I actually encode the `FramelessTransaction` variant?
2. Why did submitting `0x00` ever even partially work?

Reply:
1. Length of 1 prepended and 1 byte added for your enum variant of the tx.

2. It is an empty transaction which probably failed in the runtime than.

Reply:
> Length of 1 prepended and 1 byte added for your enum variant of the tx.

I believe that's what I've done. Length or 1 encodes as `0x04` ([citation](https://substrate.dev/docs/en/knowledgebase/advanced/codec#example-1)). My enum variant is `0x00`. That transaction fails with the error `Verification Error: Execution(ApiError(\"Could not convert parameter `tx` between node and runtime: No such variant in enum FramelessTransaction\"`

Reply:
So you had `0x0400` send to the node? (Not sure if that hex is in correct order)

Reply:
Yes, submitting `0x0400` is the example quoted above.

I've also tried submitting `0x0004` (which I think is backwards) and it behaves the same as `0x00`. This makes sense because the length prefix is 0 so the next byte is ignored.